// shadowsocks-manager/lib/plugins/flowSaver/server.js
'use strict';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

const knex = appRequire('init/knex').knex;
const manager = appRequire('services/manager');
const checkAccount = appRequire('plugins/account/checkAccount');

const add = options => {
  const name = options.name,
        host = options.host,
        port = options.port,
        password = options.password,
        method = options.method;
  var _options$scale = options.scale;
  const scale = _options$scale === undefined ? 1 : _options$scale;
  var _options$comment = options.comment;
  const comment = _options$comment === undefined ? '' : _options$comment;
  var _options$shift = options.shift;
  const shift = _options$shift === undefined ? 0 : _options$shift;

  return knex('server').insert({
    name,
    comment,
    host,
    port,
    password,
    method,
    scale,
    shift
  });
};

const del = id => {
  return knex.transaction(trx => {
    return knex('server').transacting(trx).where({ id }).delete().then(() => knex('saveFlow').transacting(trx).where({ id }).delete()).then(trx.commit).catch(trx.rollback);
  });
};

const edit = options => {
  const id = options.id,
        name = options.name,
        host = options.host,
        port = options.port,
        password = options.password,
        method = options.method;
  var _options$scale2 = options.scale;
  const scale = _options$scale2 === undefined ? 1 : _options$scale2;
  var _options$comment2 = options.comment;
  const comment = _options$comment2 === undefined ? '' : _options$comment2;
  var _options$shift2 = options.shift;
  const shift = _options$shift2 === undefined ? 0 : _options$shift2;

  checkAccount.deleteCheckAccountTimeServer(id);
  return knex('server').where({ id }).update({
    name,
    comment,
    host,
    port,
    password,
    method,
    scale,
    shift
  });
};

const list = (() => {
  var _ref = _asyncToGenerator(function* (options = {}) {
    const serverList = yield knex('server').select(['id', 'name', 'host', 'port', 'password', 'method', 'scale', 'comment', 'shift']).orderBy('name');
    if (options.status) {
      const serverStatus = [];
      const getServerStatus = function getServerStatus(server, index) {
        return manager.send({
          command: 'version'
        }, {
          host: server.host,
          port: server.port,
          password: server.password
        }).then(function (success) {
          return { status: success.version, index };
        }).catch(function (error) {
          return { status: -1, index };
        });
      };
      serverList.forEach(function (server, index) {
        serverStatus.push(getServerStatus(server, index));
      });
      const status = yield Promise.all(serverStatus);
      status.forEach(function (f) {
        serverList[f.index].status = f.status;
      });
    }
    return serverList;
  });

  return function list() {
    return _ref.apply(this, arguments);
  };
})();

exports.add = add;
exports.del = del;
exports.edit = edit;
exports.list = list;
